design/src/rfcs/rfc0022_error_context_and_compatibility.md

```diff
@@ -0,0 +1,370 @@
+
+    // Bad: Not ergonomic if additional context is added later
+    #[non_exhaustive]
+    VariantD(AnotherError),
+
+    // Bad: Exposes another library's error type in the public API,
+    // which makes upgrading or replacing that library a breaking change
+    #[non_exhaustive]
+    VariantF {
+        source: http::uri::InvalidUri
+    },
+
+    // Acceptable: Another library's error type may be in the box, and can be
+    // downcasted to, but the customer is then doing this downcasting at their
+    // own risk with no type safety guarantees that it will continue to work.
+    //
+    // Ideally, there would be a way to do opaque errors while playing
+    // nicely with fancy error reporters.
+    #[non_exhaustive]
+    VariantG {
+        source: Box<dyn Error + Send + Sync + 'static>,
+    }
+}
+
+pub struct VariantC { some_private_field: u32 }
+```
+
+When the error variants have a source, the `Error::source` method must be implemented to return that source.
+The `source` implementation _should not_ use the catch all (`_`) match arm, as this makes it easy to miss
+adding a new error variant's source at a later date.
+
+The error `Display` implementation _must not_ include the source in its output:
+
+```rust
+// Good
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::VariantA => write!(f, "variant a"),
+            Self::VariantB { some_additional_info, .. } => write!(f, "variant b ({some_additional_info})"),
+            // ... and so on
+        }
+    }
+}
+
+// Bad
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::VariantA => write!(f, "variant a"),
+            // Bad: includes the source in the `Display` output, which leads to duplicate error information
+            Self::VariantB { some_additional_info, source } => write!(f, "variant b ({some_additional_info}): {source}"),
+            // ... and so on
+        }
+    }
+}
+```
+
+### Informative error pattern
+
+Informative errors must be represented as structs. If error messaging changes based on an underlying cause, then a
+private error kind enum can be used internally for this purpose. For example:
+
+```rust
+#[derive(Debug)]
+pub struct InformativeError {
+    some_additional_info: u32,
+    source: AnotherError,
+}
+
+impl fmt::Display for InformativeError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "some informative message with {}", self.some_additional_info)
+    }
+}
+
+impl Error for InformativeError {
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        Some(&self.source)
+    }
+}
+```
+
+In general, informative errors should be referenced by variants in actionable errors since they cannot be converted
+to adaptive errors at a later date without a breaking change. This is not a hard rule, however. Use your best judgement
```

Comment:
Adaptive errors? Is that a typo or a new term?

Reply:
It's a typo.